Change withImage to tryLoadImage returning Maybe to handle failure

igagen opened the pull request:
I would like to have the ability to detect when image loading fails. The current withImage function only has a 'load' callback. I added withImage2 which also has an error callback when loading fails. This might be better called withImage', but I wasn't sure how to export the 'prime' symbol. I tried $prime, but that didn't seem to work.

paf31 commented:
Thanks! Looks good, but 

- can we implement `withImage` in terms of `withImage2`?
- is there any data in the error which can be reported by to the user?
- can we use `Either` together with a single callback in the public API? That way we can work with `ErrorT` more easily, for jobs like loading many images at once.

igagen commented:
Thanks for the feedback :) I just pushed another update addressing your comments (mostly): 

- `withImage2` now takes a single callback, using `Either` to denote error/success.
- there is now a private `withImageImpl` which is used by both `withImage` and `withImage2`.

`withImage` could also be implemented in terms of `withImage2`, but it seemed simpler to use `withImageImpl` for both.

Regarding the error callback, `onerror` doesn't provide anything in the way of status codes. It's just a generic event object without any meaningful information. It looks like you need to use an AJAX call if you want to get status codes. That kind of feels like overkill for this situation, but could be done if getting status codes back is important.

paf31 commented:
Ok, this looks much nicer IMO, but I don't think we should make an error message up if we don't have any data in the error object. If that's the case, I'd use `Maybe` instead of `Either`, and then the user can still choose to use `MaybeT` (or `ExceptT`) if they want.

igagen commented:
That makes sense. `Nothing` seems more appropriate than a manufactured error where none is given. I just pushed another update using `Maybe` instead of `Either`.

paf31 commented:
Looks great! If you could just deal with the following cosmetic points, I'll get this merged:

- Please squash your commits.
- Please update the documentation by running `pulp docs`.
- Please use consistent whitespace between declarations.
- Please rename to something like `tryLoadImage`. In fact, do you think there's much need for the old version any more? It seems we probably want to force users to deal with the `Maybe`.

Also, can you add the `canvas :: Canvas` effect? It should have been there before.

Thanks.

igagen commented:
Thanks! I updated the docs, renamed to `tryLoadImage`, added the `canvas :: Canvas` effect, cleaned up whitespace, and squashed. Let me know if I missed anything.

paf31 commented:
Did you run `pulp docs`? I think you need to run it again - the types don't match the code (see the effect rows?) and the comment is on the wrong declaration.






Sorry.... :|

igagen commented:
Oops. I forgot to rerun `pulp docs` after adding the effect. Thanks for catching that. Fixed the comment as well.

paf31 commented:
Thanks very much! And sorry for all the back and forth.

igagen commented:
Thanks for looking at this!

paf31 commented:
Thanks again, and sorry for the delay. This will go into the 2.0 release.
